staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter-config.go

```diff
@@ -0,0 +1,361 @@
+// config API objects.
+func (reqMgr *requestManager) initializeConfigController(informerFactory kubeinformers.SharedInformerFactory) {
+	reqMgr.configQueue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour), "req_mgmt_config_queue")
+	fci := informerFactory.Flowcontrol().V1alpha1()
+	pli := fci.PriorityLevelConfigurations()
+	fsi := fci.FlowSchemas()
+	reqMgr.plLister = pli.Lister()
+	reqMgr.plInformerSynced = pli.Informer().HasSynced
+	reqMgr.fsLister = fsi.Lister()
+	reqMgr.fsInformerSynced = fsi.Informer().HasSynced
+	pli.Informer().AddEventHandler(reqMgr)
+	fsi.Informer().AddEventHandler(reqMgr)
+}
+
+func (reqMgr *requestManager) triggerReload() {
+	klog.V(7).Info("triggered request-management system reloading")
+	reqMgr.configQueue.Add(0)
+}
+
+// OnAdd handles notification from an informer of object creation
+func (reqMgr *requestManager) OnAdd(obj interface{}) {
+	reqMgr.triggerReload()
+}
+
+// OnUpdate handles notification from an informer of object update
+func (reqMgr *requestManager) OnUpdate(oldObj, newObj interface{}) {
+	reqMgr.triggerReload()
+}
+
+// OnDelete handles notification from an informer of object deletion
+func (reqMgr *requestManager) OnDelete(obj interface{}) {
+	reqMgr.triggerReload()
+}
+
+func (reqMgr *requestManager) Run(stopCh <-chan struct{}) error {
+	defer reqMgr.configQueue.ShutDown()
+	klog.Info("Starting reqmgmt config controller")
+	if ok := cache.WaitForCacheSync(stopCh, reqMgr.plInformerSynced, reqMgr.fsInformerSynced); !ok {
+		return fmt.Errorf("Never achieved initial sync")
+	}
+	klog.Info("Running reqmgmt config worker")
+	wait.Until(reqMgr.runWorker, time.Second, stopCh)
+	klog.Info("Shutting down reqmgmt config worker")
+	return nil
+}
+
+func (reqMgr *requestManager) runWorker() {
+	for reqMgr.processNextWorkItem() {
+	}
+}
+
+func (reqMgr *requestManager) processNextWorkItem() bool {
+	obj, shutdown := reqMgr.configQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	func(obj interface{}) {
+		defer reqMgr.configQueue.Done(obj)
+		if !reqMgr.syncOne() {
+			reqMgr.configQueue.AddRateLimited(obj)
+		} else {
+			reqMgr.configQueue.Forget(obj)
+		}
+	}(obj)
+
+	return true
+}
+
+// syncOne attempts to sync all the config for the reqmgmt filter.  It
+// either succeeds and returns `true` or logs an error and returns
+// `false`.
+func (reqMgr *requestManager) syncOne() bool {
+	all := labels.Everything()
+	newPLs, err := reqMgr.plLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list PriorityLevelConfiguration objects: %s", err.Error())
+		return false
+	}
+	newFSs, err := reqMgr.fsLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list FlowSchema objects: %s", err.Error())
+		return false
+	}
+	reqMgr.digestConfigObjects(newPLs, newFSs)
+	return true
+}
+
+func (reqMgr *requestManager) digestConfigObjects(newPLs []*rmtypesv1a1.PriorityLevelConfiguration, newFSs []*rmtypesv1a1.FlowSchema) {
```

Comment:
OK, so here is what I would naively expect to have found here:

1) a function that copies the existing configuration
2) a function that stores a new configuration
3) 4-6 different change functions (expansion of {fs,pl}{add,update,delete}, with add & update possibly being the same function).
4) each of which gets a copy, makes a targeted change, and stores the new configuration.

Additionally I'd expect the number of workers forced to be 1, so that multiple changes don't interfere with each other.

Each of the modification functions could be unit tested.

Help me understand why instead we chose a long, delicate function which does everything?
* I think it will be hard to be sure that the unit test covers all the state transitions this is capable of performing
* I think it will be hard to maintain this in the future

Separately...
* The locking behavior might be correct, but it's not what I'll call "obviously locally" correct-- that is, you have to load large amounts of the codebase into your head to check the correctness. Given the number of maintainers that will look at and operate on this code in the future, it's much better if we can get this property. Ideally, we want the set of safe changes to this code to be obviously safe to reviewers, without those reviewers needing to look outside of the diff. We can't always get this property, but the closer the better.

I'm having trouble turning this into actionable requests. I guess here's a try:
* Split this into separate named and unit-testable modification functions (even if you continue to always execute them in one go like this)
* Make the config modification functions and the configuration publishing (i.e., the bits where locking behavior is relevant) behavior separate, such that the reader doesn't have to understand the locking behavior when reading the configuration modification, and when looking at the locking, doesn't have to know what the configuration modification was.
* unit test the two aspects separately. (There's no unit tests included for this currently, so I think this is a pretty reasonable request.)